File: solutions/java/0099.java
class Solution {
  public void recoverTree(TreeNode root) {
    inorder(root);
    int temp = x.val;
    x.val = y.val;
    y.val = temp;
  }

  private TreeNode x = null;
  private TreeNode y = null;
  private TreeNode pred = null;

  private void inorder(TreeNode root) {
    if (root == null)
      return;

    inorder(root.left);

    if (pred != null && pred.val > root.val) {
      y = root;
      if (x == null)
        x = pred;
      else
        return;
    }
    pred = root;

    inorder(root.right);
  }
}

File: solutions/cpp/0099.cpp
class Solution {
 public:
  void recoverTree(TreeNode* root) {
    vector<int> nums;

    inorder(root, nums);
    findTwoSwappedNums(nums);
    recover(root);
  }

 private:
  int x = -1;
  int y = -1;

  void inorder(TreeNode* root, vector<int>& nums) {
    if (!root) return;

    inorder(root->left, nums);
    nums.push_back(root->val);
    inorder(root->right, nums);
  }

  void findTwoSwappedNums(vector<int>& nums) {
    for (int i = 0; i + 1 < nums.size(); ++i)
      if (nums[i] > nums[i + 1]) {
        y = nums[i + 1];
        if (x == -1)
          x = nums[i];
        else
          break;
      }
  }

  void recover(TreeNode* root) {
    if (!root) return;
    if (root->val == x)
      root->val = y;
    else if (root->val == y)
      root->val = x;

    recover(root->left);
    recover(root->right);
  }
};